File: packages/contracts/src/validation/graph.ts
// validation/graph.ts

import type { NodeId, EdgeId, Workflow, Edge } from "../types";

/**
 * Adjacency list: for each node, the list of outgoing edges.
 */
export type AdjacencyList = ReadonlyMap<NodeId, readonly EdgeId[]>;

/**
 * Reverse adjacency list: for each node, the list of incoming edges.
 */
export type ReverseAdjacencyList = ReadonlyMap<NodeId, readonly EdgeId[]>;

/**
 * In-degree map: for each node, count of incoming edges.
 */
export type InDegreeMap = ReadonlyMap<NodeId, number>;

/**
 * Build adjacency list from workflow edges.
 * Maps each node to its outgoing edges.
 */
export function buildAdjacencyList(workflow: Workflow): AdjacencyList {
  const adj = new Map<NodeId, EdgeId[]>();

  // Initialize all nodes with empty arrays
  for (const nodeId of workflow.nodes.keys()) {
    adj.set(nodeId, []);
  }

  // Populate outgoing edges
  for (const [edgeId, edge] of workflow.edges) {
    const outgoing = adj.get(edge.source);
    if (outgoing) {
      outgoing.push(edgeId);
    }
  }

  return adj;
}

/**
 * Build reverse adjacency list from workflow edges.
 * Maps each node to its incoming edges.
 */
export function buildReverseAdjacencyList(
  workflow: Workflow,
): ReverseAdjacencyList {
  const rev = new Map<NodeId, EdgeId[]>();

  // Initialize all nodes with empty arrays
  for (const nodeId of workflow.nodes.keys()) {
    rev.set(nodeId, []);
  }

  // Populate incoming edges
  for (const [edgeId, edge] of workflow.edges) {
    const incoming = rev.get(edge.target);
    if (incoming) {
      incoming.push(edgeId);
    }
  }

  return rev;
}

/**
 * Compute in-degree for each node.
 */
export function computeInDegrees(workflow: Workflow): InDegreeMap {
  const degrees = new Map<NodeId, number>();

  // Initialize all nodes with zero
  for (const nodeId of workflow.nodes.keys()) {
    degrees.set(nodeId, 0);
  }

  // Count incoming edges
  for (const edge of workflow.edges.values()) {
    const current = degrees.get(edge.target) ?? 0;
    degrees.set(edge.target, current + 1);
  }

  return degrees;
}

/**
 * Find entry nodes (nodes with no incoming edges).
 */
export function findEntryNodes(workflow: Workflow): readonly NodeId[] {
  const inDegrees = computeInDegrees(workflow);
  const entries: NodeId[] = [];

  for (const [nodeId, degree] of inDegrees) {
    if (degree === 0) {
      entries.push(nodeId);
    }
  }

  return entries;
}

/**
 * Find exit nodes (nodes with no outgoing edges).
 */
export function findExitNodes(workflow: Workflow): readonly NodeId[] {
  const adj = buildAdjacencyList(workflow);
  const exits: NodeId[] = [];

  for (const [nodeId, outgoing] of adj) {
    if (outgoing.length === 0) {
      exits.push(nodeId);
    }
  }

  return exits;
}

File: apps/api/src/agentforge_api/validation/graph.py
# apps/api/src/agentforge_api/validation/graph.py

"""
Graph utilities for DAG validation.

Pure functions to transform workflow structure into forms
useful for validation algorithms.
"""

from agentforge_api.models import Node, Edge, Workflow


# Type aliases for clarity
AdjacencyList = dict[str, list[str]]      # node_id -> list of edge_ids
ReverseAdjacencyList = dict[str, list[str]]  # node_id -> list of edge_ids
InDegreeMap = dict[str, int]              # node_id -> count of incoming edges


def build_adjacency_list(workflow: Workflow) -> AdjacencyList:
    """
    Build adjacency list from workflow edges.
    Maps each node to its outgoing edge IDs.
    
    Time: O(V + E)
    Space: O(V + E)
    """
    adj: AdjacencyList = {}
    
    # Initialize all nodes with empty lists
    for node in workflow.nodes:
        adj[node.id] = []
    
    # Populate outgoing edges
    edge_map = workflow.get_edge_map()
    for edge in workflow.edges:
        if edge.source in adj:
            adj[edge.source].append(edge.id)
    
    return adj


def build_reverse_adjacency_list(workflow: Workflow) -> ReverseAdjacencyList:
    """
    Build reverse adjacency list from workflow edges.
    Maps each node to its incoming edge IDs.
    
    Time: O(V + E)
    Space: O(V + E)
    """
    rev: ReverseAdjacencyList = {}
    
    # Initialize all nodes with empty lists
    for node in workflow.nodes:
        rev[node.id] = []
    
    # Populate incoming edges
    for edge in workflow.edges:
        if edge.target in rev:
            rev[edge.target].append(edge.id)
    
    return rev


def compute_in_degrees(workflow: Workflow) -> InDegreeMap:
    """
    Compute in-degree for each node.
    
    Time: O(V + E)
    Space: O(V)
    """
    degrees: InDegreeMap = {}
    
    # Initialize all nodes with zero
    for node in workflow.nodes:
        degrees[node.id] = 0
    
    # Count incoming edges
    for edge in workflow.edges:
        if edge.target in degrees:
            degrees[edge.target] += 1
    
    return degrees


def find_entry_nodes(workflow: Workflow) -> list[str]:
    """
    Find entry nodes (nodes with no incoming edges).
    
    Time: O(V + E)
    Space: O(V)
    """
    in_degrees = compute_in_degrees(workflow)
    
    return [
        node_id
        for node_id, degree in in_degrees.items()
        if degree == 0
    ]


def find_exit_nodes(workflow: Workflow) -> list[str]:
    """
    Find exit nodes (nodes with no outgoing edges).
    
    Time: O(V + E)
    Space: O(V)
    """
    adj = build_adjacency_list(workflow)
    
    return [
        node_id
        for node_id, outgoing in adj.items()
        if len(outgoing) == 0
    ]